### PR TITLE
Migrate Instrumentation and Visibility to doc-scope service.

### PR DESCRIPTION
**Pull request by mkhatib**
tracking bug: #5565

### CODE REVIEW

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 1]
```diff
@@ -15,28 +15,26 @@
...
 import {installActivityService} from './activity-impl';
-import {installVisibilityService} from './visibility-impl';
 import {isArray, isObject} from '../../../src/types';
 import {sendRequest, sendRequestUsingIframe} from './transport';
 import {urlReplacementsForDoc} from '../../../src/url-replacements';
 import {userNotificationManagerFor} from '../../../src/user-notification';
 import {cryptoFor} from '../../../src/crypto';
 import {xhrFor} from '../../../src/xhr';
 import {toggle} from '../../../src/style';
+import './visibility-impl';
```
Comment (mkhatib) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Let me know if we should still wrap the `AMP.registerServiceForDoc` calls with a function and explicitly call it here.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 2]
```diff
@@ -177,11 +175,11 @@ export class AmpAnalytics extends AMP.BaseElement {
             trigger['selector'] = this.expandTemplate_(trigger['selector'],
                 trigger, /* arg*/ undefined, /* arg */ undefined,
                 /* arg*/ false);
-            addListener(this.win, trigger, this.handleEvent_.bind(this,
+            addListener(this.win.document, trigger, this.handleEvent_.bind(this,
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: This should just be now `getAmpDoc()`, right? Why delay migration here? Rather, actually, it'd be a lot better if we called `instrumentation.addEventListener` directly here. I don't see a point in resolving it downstream.

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 3]
```diff
@@ -39,15 +39,15 @@ const VARIABLE_DATA_ATTRIBUTE_KEY = /^vars(.+)/;
 let AnalyticsEventListenerDef;
 
 /**
- * @param {!Window} window Window object to listen on.
+ * @param {!Node|!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: `nodeOrDoc`

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 4]
```diff
@@ -517,11 +519,19 @@ export class InstrumentationService {
 }
 
 /**
- * @param {!Window} window
+ * @param {!Node|!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: `nodeOrDoc`

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Yeah, this is an annoyance. This is why it's better to keep all registrations in the root module for extension - in other words here. A set of `AMP.registerServiceForDoc` is a bit unsightly, but a better alternative then orphaned `imports`.

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 5]
```diff
@@ -30,6 +30,8 @@ import {installCidService,} from
 import {urlReplacementsForDoc} from '../../../../src/url-replacements';
 import * as sinon from 'sinon';
 
+import {AmpDocSingle} from '../../../../src/service/ampdoc-impl';
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: Any way to switch to `describes.fakeWin` or `realWin` - they pre-created ampdoc environment for you.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 6]
```diff
@@ -161,7 +161,7 @@ export function getElement(selector, el, selectionMethod) {
   } else if (selectionMethod == 'scope') {
     return el.parentElement.querySelector(selector);
   } else if (selector[0] == '#') {
-    return el.ownerDocument.getElementById(selector.slice(1));
+    return ampdoc.getElementById(selector.slice(1));
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: A note for the related code. Five lines above (I can't comment there) we do `closest` selectors - we should restrict the result to be contained by `ampdoc`

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 7]
```diff
@@ -187,9 +187,13 @@ let VisibilityListenerDef;
  */
 export class Visibility {
 
-  /** @param {!Window} win */
-  constructor(win) {
-    this.win_ = win;
+  /** @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+    /** @private @const {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc */
+    this.ampdoc_ = ampdoc;
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Canonically, it should be a public instance var.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 8]
```diff
@@ -187,9 +187,13 @@ let VisibilityListenerDef;
...
+
+    /** @private @const {!Window} */
+    this.win_ = this.ampdoc_.win;
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: I'd prefer to remove `this.win_` - it should be harder to access global.

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 9]
```diff
@@ -531,10 +536,20 @@ export class Visibility {
...
  */
-export function installVisibilityService(win) {
-  return fromClass(win, 'visibility', Visibility);
+function installVisibilityServiceForDoc(ampdoc) {
+  return fromClassForDoc(ampdoc, 'visibility', Visibility);
 };
+
+
+// Register doc-service factory.
+AMP.registerServiceForDoc(
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Do we really need this? We can just create this instance from `Instrumentation`. I don't think it exists outside of this scope. If it does - please explain.

File: src/dom.js [hunk 10]
```diff
@@ -68,7 +68,7 @@ export function waitForChild(parent, checkFunc, callback) {
  * Waits for document's body to be available.
  * Will be deprecated soon; use {@link AmpDoc#whenBodyAvailable} or
  * @{link DocumentState#onBodyAvailable} instead.
- * @param {!Document} doc
+ * @param {!Document|!ShadowRoot} doc
```
Comment (dvoytenko) [inline src/dom.js]: Please remove.

File: src/dom.js [hunk 11]
```diff
@@ -78,7 +78,7 @@ export function waitForBody(doc, callback) {
 
 /**
  * Waits for document's body to be available.
- * @param {!Document} doc
+ * @param {!Document|!ShadowRoot} doc
```
Comment (dvoytenko) [inline src/dom.js]: Please remove.

File: src/service.js [hunk 12]
```diff
@@ -268,7 +268,7 @@ export function getParentWindowFrameElement(node, topWin) {
  * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
  * @return {!./service/ampdoc-impl.AmpDoc}
  */
-function getAmpdoc(nodeOrDoc) {
+export function getAmpdoc(nodeOrDoc) {
```
Comment (dvoytenko) [inline src/service.js]: Shouldn't be exported.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 9 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/visibility-impl.js]: You mean dropping registering visibility service all together? Yeah it seems its only used in instrumentation so we can drop it. Is that what you meant?

File: src/service.js (hunk 12 above)
Comment (mkhatib) [inline src/service.js]: I am using this in `element-service.js` why this shouldn't be exported?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 6 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Is that for the `closest` polyfill or for both? Do we just pass the ampdoc to the call and stop closest ancestor traversal when we reach it?

File: src/dom.js (hunk 11 above)
Comment (mkhatib) [inline src/dom.js]: Our type check complains about this. Do we call `waitForBodyPromise` with a document instead?

```
src/element-service.js:144: ERROR - actual parameter 1 of module$src$dom.waitForBodyPromise does not match formal parameter
found   : (Document|ShadowRoot)
required: Document
    return dom.waitForBodyPromise(ampdoc.getRootNode()).then(() => {
```

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: I'd just do it here. And I'd simply check if ampdoc contains node and if it doesn't - force the value to `null`.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 9 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Yes. Just create it there once and use it w/o lookup.

File: src/element-service.js [hunk 13]
```diff
@@ -85,3 +91,61 @@ function isElementScheduled(win, elementName) {
...
+  }
+  // Microtask is necessary to ensure that window.ampExtendedElements has been
+  // initialized.
+  return Promise.resolve().then(() => {
+    if (isElementScheduled(ampdoc.win, providedByElement)) {
+      return getServicePromiseForDoc(nodeOrDoc, id);
+    }
+    // Wait for HEAD to fully form before denying access to the service.
+    // TODO(mk): DOES THIS NEED TO BE CHANGED?
+    return dom.waitForBodyPromise(ampdoc.getRootNode()).then(() => {
```
Comment (dvoytenko) [inline src/element-service.js]: So, as described elsewhere as well: `waitForBody` is not the right call here. Instead we should just use `ampdoc.whenBodyAvailable`. And yes, you need to wait for body here.

File: src/service.js (hunk 12 above)
Comment (dvoytenko) [inline src/service.js]: Hmm. This is a private method. But ok. Let's expose for now.

File: src/dom.js (hunk 10 above)
Comment (mkhatib) [inline src/dom.js]: Done

File: src/dom.js (hunk 11 above)
Comment (mkhatib) [inline src/dom.js]: Done

File: src/service.js (hunk 12 above)
Comment (mkhatib) [inline src/service.js]: Ack

File: src/element-service.js (hunk 13 above)
Comment (mkhatib) [inline src/element-service.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 3 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 8 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Done

**Comment by mkhatib** [comment]
Will work on updating tests and try to use `describes` now but all other comments should be addressed

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 14]
```diff
@@ -39,15 +39,15 @@ const VARIABLE_DATA_ATTRIBUTE_KEY = /^vars(.+)/;
...
+ * @param {!Node|!../../../src/service/ampdoc-impl.AmpDoc} nodeOrDoc
  * @param {!JSONType} config Configuration for instrumentation.
  * @param {!AnalyticsEventListenerDef} listener Callback to call when the event
  *  fires.
  * @param {!Element} analyticsElement The element associated with the
  *  config.
  */
-export function addListener(window, config, listener, analyticsElement) {
-  return instrumentationServiceFor(window).addListener(config, listener,
+export function addListener(nodeOrDoc, config, listener, analyticsElement) {
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/instrumentation.js]: is this function used anywhere now? Maybe remove it?

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js (hunk 5 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: I'd rather not delay this PR anymore, so filed a bug to track moving to describes. For now, I'll go ahead and fix the tests and rebase.

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 14 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

**Comment by mkhatib** [comment]
Alright, updated tests 👀 PTAL

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 15]
```diff
@@ -32,11 +31,17 @@ import {cryptoFor} from '../../../src/crypto';
 import {xhrFor} from '../../../src/xhr';
 import {toggle} from '../../../src/style';
 
+// Register doc-service factory.
+AMP.registerServiceForDoc(
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: The best option is to register the service like so:
```
AMP.registerServiceForDoc('amp-analytics-instrumentation', InstrumentationService)
```

And change `instrumentationServiceForDoc` to return the service and not create a new one.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 16]
```diff
@@ -175,17 +180,19 @@ export class AmpAnalytics extends AMP.BaseElement {
             return;
           }
 
+          const instrumentation_ = instrumentationServiceForDoc(
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Resolve the instrumentation service once in the `buildCallback` and use private var reference.

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 17]
```diff
@@ -519,19 +509,18 @@ export class InstrumentationService {
...
 
 /**
- * @param {!Window} window
+ * @param {!Node|!../../../src/service/ampdoc-impl.AmpDoc} nodeOrDoc
  * @return {!InstrumentationService}
  */
-export function instrumentationServiceFor(window) {
-  return fromClass(window, 'amp-analytics-instrumentation',
+export function instrumentationServiceForDoc(nodeOrDoc) {
+  return fromClassForDoc(nodeOrDoc, 'amp-analytics-instrumentation',
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/instrumentation.js]: As mentioned above, use `getExistingServiceForDoc` instead. It's very dangerous to install a service twice.

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: sg

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 18]
```diff
@@ -155,27 +154,31 @@ export function isVisibilitySpecValid(config) {
...
+    const elWin = ampdoc.win;
     const parentEl = elWin.frameElement && elWin.frameElement.parentElement;
     if (parentEl) {
       return closestBySelector(parentEl, '.-amp-element');
     }
   }
 
   if (selectionMethod == 'closest') {
+    // Restrict result to be contained by ampdoc.
+    if (!ampdoc.contains(el)) {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: I don't think this is what you need to do, unless I'm missing something. The input `el` is definitely in the ampdoc. However, the result of `closestByTag` may not be. Consider this:
```
result
  div
    - ampdoc-boundary - 
          container
             el
```

You are looking at `el` and then querying `closestByTag(el)` which yields `result`. So what just happened is that you traversed the ampdoc boundary by reaching to `result`. Thus all we need to do is to check that `ampdoc.contains(result)`.

File: src/element-service.js [hunk 19]
```diff
@@ -85,3 +91,60 @@ function isElementScheduled(win, elementName) {
...
+/**
+ * Same as getElementService but produces null if the given element is not
+ * actually available on the current page.
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
+ * @param {string} id of the service.
+ * @param {string} providedByElement Name of the custom element that provides
+ *     the implementation of this service.
+ * @return {!Promise<?Object>}
+ */
+export function getElementServiceIfAvailableForDoc(
```
Comment (dvoytenko) [inline src/element-service.js]: Not sure I saw tests for these two methods.

Comment (mkhatib) [inline src/element-service.js]: There wasn't any tests for this file at all, should I create one or file an issue to follow up on it?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 18 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Fixed.

File: extensions/amp-analytics/0.1/instrumentation.js (hunk 17 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/instrumentation.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 16 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 15 above)
Comment (mkhatib) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done

File: examples/forms.amp.html [hunk 20]
```diff
@@ -166,7 +166,7 @@
 </form>
 
 <h4>Search (GET xhr same origin)</h4>
-<form method="get" action="/form/search-html/get" action-xhr="/form/search-json/get" target="_blank">
+<form method="get" action="/form/search-html/get" action-xhr="https://www.truthfinder.com/search/" target="_blank">
```
Comment (mkhatib) [inline examples/forms.amp.html]: Done

File: src/element-service.js (hunk 19 above)
Comment (dvoytenko) [inline src/element-service.js]: The tests are in the `test-custom-element.js`. Looks like they didn't follow migration with the new file. Let's add them there.

Comment (mkhatib) [inline src/element-service.js]: Added tests 😄

**Comment by mkhatib** [comment]
PTAL 👀

**Comment by mkhatib** [comment]
Going ahead and merging this. I was unable to test the use-case for making sure clicks are surfaced from the shadow dom because of other parts of analytics are still throwing the `no ampdoc found` so will work on that next and getting the rest of AMP analytics to work. 

cc/ @avimehta @dvoytenko

**Pull request merged**